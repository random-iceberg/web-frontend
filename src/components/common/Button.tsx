import React from "react";

/**
 * Button style variants.
 * @typedef {'primary' | 'secondary' | 'danger' | 'link'} ButtonVariant
 */
type ButtonVariant = "primary" | "secondary" | "danger" | "link";

/**
 * Button size options.
 * @typedef {'sm' | 'md' | 'lg'} ButtonSize
 */
type ButtonSize = "sm" | "md" | "lg";

/**
 * Props for the Button component. Extends standard HTML button attributes.
 * @typedef {object} ButtonProps
 * @property {ButtonVariant} [variant='primary'] - The style variant of the button.
 * @property {ButtonSize} [size='md'] - The size of the button.
 * @property {boolean} [fullWidth=false] - Whether the button should take up the full width of its container.
 * @property {React.ReactNode} children - The content inside the button.
 * @property {string} [className] - Additional CSS classes to apply.
 */
type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  variant?: ButtonVariant;
  size?: ButtonSize;
  fullWidth?: boolean;
  children: React.ReactNode;
};

/**
 * A versatile button component with different variants and sizes.
 * It accepts all standard HTML button attributes like `onClick`, `disabled`, `type`, etc.
 *
 * @component
 * @param {ButtonProps} props - Props for the Button component.
 * @example
 * // Primary Button
 * <Button onClick={() => console.log('Clicked!')}>Submit</Button>
 *
 * @example
 * // Secondary Button (Small)
 * <Button variant="secondary" size="sm" onClick={handleCancel}>Cancel</Button>
 *
 * @example
 * // Danger Button (Full Width, Disabled)
 * <Button variant="danger" fullWidth disabled={isDeleting}>Delete</Button>
 *
 * @example
 * // Link style button
 * <Button variant="link" onClick={showDetails}>View Details</Button>
 */
const Button: React.FC<ButtonProps> = ({
  variant = "primary",
  size = "md",
  fullWidth = false,
  className = "",
  children,
  ...props // Pass down other button attributes like type, onClick, disabled
}) => {
  // Base styles
  const base =
    "inline-flex items-center justify-center font-medium rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed transition ease-in-out duration-150";

  // Size styles
  const sizes: Record<ButtonSize, string> = {
    sm: "px-3 py-1.5 text-xs",
    md: "px-4 py-2 text-sm",
    lg: "px-6 py-3 text-base",
  };

  // Variant styles
  const variants: Record<ButtonVariant, string> = {
    primary:
      "bg-blue-600 text-white hover:bg-blue-700 focus:ring-blue-500 border border-transparent",
    secondary:
      "bg-white text-gray-700 border border-gray-300 hover:bg-gray-50 focus:ring-indigo-500", // Adjusted focus ring for visibility
    danger:
      "bg-red-600 text-white hover:bg-red-700 focus:ring-red-500 border border-transparent",
    link: "text-blue-600 hover:text-blue-800 underline focus:ring-blue-500 shadow-none border-none bg-transparent p-0", // Link style
  };

  const width = fullWidth ? "w-full" : "";

  const combinedClassName =
    `${base} ${sizes[size]} ${variants[variant]} ${width} ${className}`.trim();

  return (
    <button className={combinedClassName} {...props}>
      {children}
    </button>
  );
};

export default Button;
