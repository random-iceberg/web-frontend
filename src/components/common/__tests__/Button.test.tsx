import React from "react";
import { render } from "@testing-library/react";
import Button from "../Button";

describe("Button Component", () => {
  it("renders primary button correctly", () => {
    const { asFragment } = render(<Button>Primary</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders secondary button correctly", () => {
    const { asFragment } = render(
      <Button variant="secondary">Secondary</Button>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders danger button correctly", () => {
    const { asFragment } = render(<Button variant="danger">Danger</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders link button correctly", () => {
    const { asFragment } = render(<Button variant="link">Link</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders small button correctly", () => {
    const { asFragment } = render(<Button size="sm">Small</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders large button correctly", () => {
    const { asFragment } = render(<Button size="lg">Large</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders fullWidth button correctly", () => {
    const { asFragment } = render(<Button fullWidth>Full Width</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders disabled button correctly", () => {
    const { asFragment } = render(<Button disabled>Disabled</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with additional className", () => {
    const { asFragment } = render(
      <Button className="my-custom-class">Custom Class</Button>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with onClick handler (structure check)", () => {
    const mockOnClick = jest.fn();
    const { asFragment } = render(
      <Button onClick={mockOnClick}>Click Me</Button>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
