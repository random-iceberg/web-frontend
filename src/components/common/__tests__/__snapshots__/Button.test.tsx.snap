// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component renders danger button correctly 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center font-medium rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed transition ease-in-out duration-150 px-4 py-2 text-sm bg-red-600 text-white hover:bg-red-700 focus:ring-red-500 border border-transparent"
  >
    Danger
  </button>
</DocumentFragment>
`;

exports[`Button Component renders disabled button correctly 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center font-medium rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed transition ease-in-out duration-150 px-4 py-2 text-sm bg-primary text-white hover:bg-accent focus:ring-primary border border-transparent"
    disabled=""
  >
    Disabled
  </button>
</DocumentFragment>
`;

exports[`Button Component renders fullWidth button correctly 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center font-medium rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed transition ease-in-out duration-150 px-4 py-2 text-sm bg-primary text-white hover:bg-accent focus:ring-primary border border-transparent w-full"
  >
    Full Width
  </button>
</DocumentFragment>
`;

exports[`Button Component renders large button correctly 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center font-medium rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed transition ease-in-out duration-150 px-6 py-3 text-base bg-primary text-white hover:bg-accent focus:ring-primary border border-transparent"
  >
    Large
  </button>
</DocumentFragment>
`;

exports[`Button Component renders link button correctly 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center font-medium rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed transition ease-in-out duration-150 px-4 py-2 text-sm text-primary hover:text-accent underline focus:ring-primary shadow-none border-none bg-transparent p-0"
  >
    Link
  </button>
</DocumentFragment>
`;

exports[`Button Component renders primary button correctly 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center font-medium rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed transition ease-in-out duration-150 px-4 py-2 text-sm bg-primary text-white hover:bg-accent focus:ring-primary border border-transparent"
  >
    Primary
  </button>
</DocumentFragment>
`;

exports[`Button Component renders secondary button correctly 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center font-medium rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed transition ease-in-out duration-150 px-4 py-2 text-sm bg-white text-gray-700 border border-gray-300 hover:bg-gray-100 focus:ring-secondary"
  >
    Secondary
  </button>
</DocumentFragment>
`;

exports[`Button Component renders small button correctly 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center font-medium rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed transition ease-in-out duration-150 px-3 py-1.5 text-xs bg-primary text-white hover:bg-accent focus:ring-primary border border-transparent"
  >
    Small
  </button>
</DocumentFragment>
`;

exports[`Button Component renders with additional className 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center font-medium rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed transition ease-in-out duration-150 px-4 py-2 text-sm bg-primary text-white hover:bg-accent focus:ring-primary border border-transparent  my-custom-class"
  >
    Custom Class
  </button>
</DocumentFragment>
`;

exports[`Button Component renders with onClick handler (structure check) 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center font-medium rounded-md shadow-sm focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed transition ease-in-out duration-150 px-4 py-2 text-sm bg-primary text-white hover:bg-accent focus:ring-primary border border-transparent"
  >
    Click Me
  </button>
</DocumentFragment>
`;
